cli-plugins/socket/socket.go

```diff
@@ -7,16 +7,98 @@ import (
 	"io"
 	"net"
 	"os"
+	"sync"
+	"time"
 )
 
 // EnvKey represents the well-known environment variable used to pass the plugin being
 // executed the socket name it should listen on to coordinate with the host CLI.
 const EnvKey = "DOCKER_CLI_PLUGIN_SOCKET"
 
+// UnixConnSafe is a wrapper around a UnixConn
+// that makes it concurrency-safe.
+type UnixConnSafe struct {
+	UnixConn   *net.UnixConn
+	SocketLock *sync.Mutex
+}
+
+var _ net.Conn = (*UnixConnSafe)(nil)
+
+func NewUnixConnSafe() *UnixConnSafe {
+	return &UnixConnSafe{
+		UnixConn:   &net.UnixConn{},
+		SocketLock: &sync.Mutex{},
```

Comment:
Where possible, it's good to keep mutexes "internal" nobody outside of "UnixConnSafe" itself should really have to deal with locking/unlocking.